neuraxle/base.py

```diff
@@ -82,6 +82,42 @@ def append(self, current_id, data_input, expected_output):
         self.data_inputs.append(data_input)
         self.expected_outputs.append(expected_output)
 
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            start = key.start
+            stop = key.stop
+            step = key.step
+            if step is not None or (start is None and stop is None):
+                raise KeyError("Invalid range: '{}'.".format(key))
+
+            if start is None:
+                return DataContainer(
+                    data_inputs=self.data_inputs[:stop],
+                    current_ids=self.current_ids[:stop],
+                    expected_outputs=self.expected_outputs[:stop],
+                )
+            elif stop is None:
+                return DataContainer(
+                    data_inputs=self.data_inputs[start:],
+                    current_ids=self.current_ids[start:],
+                    expected_outputs=self.expected_outputs[start:],
+                )
+            else:
+                return DataContainer(
+                    data_inputs=self.data_inputs[start:stop],
+                    current_ids=self.current_ids[start:stop],
+                    expected_outputs=self.expected_outputs[start:stop],
+                )
+        else:
+            return self.current_ids[key], self.data_inputs[key], self.expected_outputs[key]
+
+    def __reversed__(self):
+        return DataContainer(
+            data_inputs=list(reversed(self.data_inputs)),
+            current_ids=list(reversed(self.current_ids)),
+            expected_outputs=list(reversed(self.expected_outputs)),
```

Comment:
Don't cast as list here! `data_inputs` and/or `expected_outputs` might be a generator (e.g.: a lazy load). Reversing a step should not require loading all the database at once. 

Perhaps only maybe the `current_ids` can be casted as list.

Reply:
However, __get_item__ needs to be a list to be subscriptable. Also, if we get to the point where we need to perform a __get_item__ operation, the data should already have been loaded.

Reply:
this would also cause the following error if we try to reverse the data container again if we keep the iterator type : 'list_iterator' object is not reversible. 

can we please do this in another pr. if we really need a bidirectional iterator, then we'll do it.

Reply:
won't fix

Reply:
Could the iterator be reversible? 
I'd prefer not to merge this code.

Reply:
@guillaume-chevalier cancelling this pr. this will remain in the history if we ever need this.